// domain-search-mcp/dist/api/routes.js
"use strict";
/**
 * REST API Routes for Tool Execution
 *
 * Provides HTTP endpoints for each MCP tool.
 * Used by ChatGPT Actions and other REST API clients.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApiRouter = createApiRouter;
const express_1 = require("express");
const index_js_1 = require("../tools/index.js");
const errors_js_1 = require("../utils/errors.js");
// Map tool names to their executors
const TOOL_EXECUTORS = {
    search_domain: index_js_1.executeSearchDomain,
    bulk_search: index_js_1.executeBulkSearch,
    compare_registrars: index_js_1.executeCompareRegistrars,
    suggest_domains: index_js_1.executeSuggestDomains,
    suggest_domains_smart: index_js_1.executeSuggestDomainsSmart,
    tld_info: index_js_1.executeTldInfo,
    check_socials: index_js_1.executeCheckSocials,
    analyze_project: index_js_1.executeAnalyzeProject,
    hunt_domains: index_js_1.executeHuntDomains,
    expiring_domains: index_js_1.executeExpiringDomains,
};
/**
 * Create REST API router for tools.
 */
function createApiRouter() {
    const router = (0, express_1.Router)();
    // Generic tool execution handler
    const handleToolExecution = (toolName) => {
        return async (req, res, next) => {
            const executor = TOOL_EXECUTORS[toolName];
            if (!executor) {
                res.status(404).json({
                    success: false,
                    error: {
                        code: 'UNKNOWN_TOOL',
                        message: `Tool "${toolName}" not found`,
                    },
                });
                return;
            }
            try {
                const args = req.body || {};
                const result = await executor(args);
                res.json({
                    success: true,
                    data: result,
                });
            }
            catch (error) {
                const wrapped = (0, errors_js_1.wrapError)(error);
                // Determine HTTP status from error code
                let status = 500;
                if (wrapped.code === 'VALIDATION_ERROR' || wrapped.code === 'INVALID_INPUT') {
                    status = 400;
                }
                else if (wrapped.code === 'RATE_LIMIT_EXCEEDED') {
                    status = 429;
                }
                else if (wrapped.code === 'NOT_FOUND') {
                    status = 404;
                }
                res.status(status).json({
                    success: false,
                    error: {
                        code: wrapped.code,
                        message: wrapped.userMessage,
                        retryable: wrapped.retryable,
                    },
                });
            }
        };
    };
    // Register routes for all tools
    router.post('/tools/search_domain', handleToolExecution('search_domain'));
    router.post('/tools/bulk_search', handleToolExecution('bulk_search'));
    router.post('/tools/compare_registrars', handleToolExecution('compare_registrars'));
    router.post('/tools/suggest_domains', handleToolExecution('suggest_domains'));
    router.post('/tools/suggest_domains_smart', handleToolExecution('suggest_domains_smart'));
    router.post('/tools/tld_info', handleToolExecution('tld_info'));
    router.post('/tools/check_socials', handleToolExecution('check_socials'));
    router.post('/tools/analyze_project', handleToolExecution('analyze_project'));
    router.post('/tools/hunt_domains', handleToolExecution('hunt_domains'));
    router.post('/tools/expiring_domains', handleToolExecution('expiring_domains'));
    return router;
}
