// domain-search-mcp/dist/services/vps-suggest.js
"use strict";
/**
 * VPS Suggest Service
 *
 * Calls VPS backend for AI-powered domain suggestions.
 * All API keys (Together.ai) are stored on VPS, not in MCP.
 * This allows npm package users to get AI suggestions without any API keys.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVpsSuggestConfigured = isVpsSuggestConfigured;
exports.fetchVpsSuggestions = fetchVpsSuggestions;
exports.checkVpsSuggestHealth = checkVpsSuggestHealth;
const config_js_1 = require("../config.js");
/**
 * Check if VPS suggest API is configured (PRICING_API_BASE_URL is set)
 */
function isVpsSuggestConfigured() {
    return config_js_1.config.pricingApi.enabled && !!config_js_1.config.pricingApi.baseUrl;
}
/**
 * Fetch AI-powered domain suggestions from VPS backend
 */
async function fetchVpsSuggestions(request) {
    if (!config_js_1.config.pricingApi.enabled || !config_js_1.config.pricingApi.baseUrl) {
        return null;
    }
    const baseUrl = config_js_1.config.pricingApi.baseUrl.replace(/\/$/, '');
    const url = `${baseUrl}/api/suggest`;
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 30000); // 30s timeout
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                ...(config_js_1.config.pricingApi.token
                    ? { Authorization: `Bearer ${config_js_1.config.pricingApi.token}` }
                    : {}),
            },
            body: JSON.stringify(request),
            signal: controller.signal,
        });
        clearTimeout(timeoutId);
        if (!response.ok) {
            console.error(`VPS suggest error: ${response.status}`);
            return null;
        }
        const data = (await response.json());
        return data;
    }
    catch (error) {
        if (error instanceof Error && error.name === 'AbortError') {
            console.error('VPS suggest request timed out');
        }
        else {
            console.error('VPS suggest request failed:', error);
        }
        return null;
    }
}
/**
 * Check VPS suggest API health
 */
async function checkVpsSuggestHealth() {
    if (!config_js_1.config.pricingApi.enabled || !config_js_1.config.pricingApi.baseUrl) {
        return { ok: false };
    }
    const baseUrl = config_js_1.config.pricingApi.baseUrl.replace(/\/$/, '');
    const url = `${baseUrl}/api/suggest/health`;
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 5000);
        const response = await fetch(url, {
            signal: controller.signal,
        });
        clearTimeout(timeoutId);
        if (!response.ok) {
            return { ok: false };
        }
        return (await response.json());
    }
    catch {
        return { ok: false };
    }
}
