// domain-search-mcp/dist/tools/compare_registrars.js
"use strict";
/**
 * compare_registrars Tool - Price Comparison.
 *
 * Compare pricing across multiple registrars for a specific domain.
 * Helps find the best deal.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.compareRegistrarsTool = exports.compareRegistrarsSchema = void 0;
exports.executeCompareRegistrars = executeCompareRegistrars;
const zod_1 = require("zod");
const domain_search_js_1 = require("../services/domain-search.js");
const validators_js_1 = require("../utils/validators.js");
const errors_js_1 = require("../utils/errors.js");
/**
 * Input schema for compare_registrars.
 */
exports.compareRegistrarsSchema = zod_1.z.object({
    domain: zod_1.z
        .string()
        .min(1)
        .describe("The domain name to compare (e.g., 'vibecoding')."),
    tld: zod_1.z
        .string()
        .describe("The TLD extension (e.g., 'com', 'io')."),
    registrars: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe("Registrars to compare (e.g., ['porkbun']). Defaults to all available."),
});
/**
 * Tool definition for MCP.
 */
exports.compareRegistrarsTool = {
    name: 'compare_registrars',
    description: `Compare domain pricing across multiple registrars.

Checks the same domain at different registrars to find:
- Best first year price
- Best renewal price
- Overall recommendation

Uses the Pricing API when configured; otherwise falls back to BYOK registrars.

Returns pricing comparison and a recommendation.

Example:
- compare_registrars("vibecoding", "com") → compares Porkbun (and any configured BYOK registrars)`,
    inputSchema: {
        type: 'object',
        properties: {
            domain: {
                type: 'string',
                description: "The domain name to compare (without extension).",
            },
            tld: {
                type: 'string',
                description: "The TLD extension (e.g., 'com', 'io').",
            },
            registrars: {
                type: 'array',
                items: { type: 'string' },
                description: "Registrars to compare. Defaults to ['porkbun'].",
            },
        },
        required: ['domain', 'tld'],
    },
};
/**
 * Execute the compare_registrars tool.
 */
async function executeCompareRegistrars(input) {
    try {
        const { domain, tld, registrars } = exports.compareRegistrarsSchema.parse(input);
        const normalizedDomain = (0, validators_js_1.validateDomainName)(domain);
        const normalizedTld = (0, validators_js_1.validateTld)(tld);
        const fullDomain = `${normalizedDomain}.${normalizedTld}`;
        const result = await (0, domain_search_js_1.compareRegistrars)(normalizedDomain, normalizedTld, registrars);
        const insights = [];
        // Generate insights
        if (result.best_first_year && result.best_renewal) {
            if (result.best_first_year.registrar === result.best_renewal.registrar) {
                insights.push(`✅ ${result.best_first_year.registrar} wins on both first year and renewal`);
            }
            else {
                insights.push(`💡 Split strategy: ${result.best_first_year.registrar} for year 1, consider transfer to ${result.best_renewal.registrar} later`);
            }
        }
        // Check for privacy inclusion
        const withPrivacy = result.comparisons.filter((r) => r.privacy_included);
        if (withPrivacy.length > 0) {
            insights.push(`🔒 ${withPrivacy.map((r) => r.registrar).join(', ')} include free WHOIS privacy`);
        }
        // Premium warning
        const premiums = result.comparisons.filter((r) => r.premium);
        if (premiums.length > 0) {
            insights.push(`⚠️ This is a premium domain at: ${premiums.map((r) => r.registrar).join(', ')}`);
        }
        if (result.best_first_year?.registrar) {
            const winner = result.comparisons.find((r) => r.registrar === result.best_first_year?.registrar);
            if (winner?.price_check_url) {
                insights.push(`Verify pricing for ${fullDomain}: ${winner.price_check_url}`);
            }
        }
        insights.push('⚠️ Prices can change. Verify at registrar checkout links before purchase.');
        return {
            domain: fullDomain,
            what_happened: `Compared pricing across ${result.comparisons.length} registrars`,
            comparison_count: result.comparisons.length,
            comparisons: result.comparisons,
            best_first_year: result.best_first_year
                ? { ...result.best_first_year, currency: 'USD' }
                : null,
            best_renewal: result.best_renewal
                ? { ...result.best_renewal, currency: 'USD' }
                : null,
            recommendation: result.recommendation,
            insights,
        };
    }
    catch (error) {
        throw (0, errors_js_1.wrapError)(error);
    }
}
