// domain-search-mcp/dist/tools/search_domain.js
"use strict";
/**
 * search_domain Tool - Primary Domain Search.
 *
 * This is the main entry point for domain availability checks.
 * Designed for magic first moment: works with zero configuration.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchDomainTool = exports.searchDomainSchema = void 0;
exports.executeSearchDomain = executeSearchDomain;
const zod_1 = require("zod");
const domain_search_js_1 = require("../services/domain-search.js");
const errors_js_1 = require("../utils/errors.js");
/**
 * Input schema for search_domain.
 */
exports.searchDomainSchema = zod_1.z.object({
    domain_name: zod_1.z
        .string()
        .min(1)
        .max(63)
        .describe("The domain name you want to check (e.g., 'vibecoding', 'myapp'). Don't include the extension (.com, .io, etc.)"),
    tlds: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe("TLD extensions to check (e.g., ['com', 'io', 'dev']). Defaults to ['com', 'io', 'dev'] if not specified."),
    registrars: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe("Optional: specific registrars to check (BYOK only; ignored when Pricing API is configured)."),
});
/**
 * Tool definition for MCP.
 */
exports.searchDomainTool = {
    name: 'search_domain',
    description: `Search for domain availability and pricing across multiple TLDs.

Returns:
- Availability status for each domain
- Pricing (first year and renewal) when Pricing API is configured
- Whether WHOIS privacy is included
- Human-readable insights and next steps

Examples:
- search_domain("vibecoding") → checks vibecoding.com, .io, .dev
- search_domain("myapp", ["com", "io"]) → checks specific TLDs`,
    inputSchema: {
        type: 'object',
        properties: {
            domain_name: {
                type: 'string',
                description: "The domain name to search (e.g., 'vibecoding'). No extension needed.",
            },
            tlds: {
                type: 'array',
                items: { type: 'string' },
                description: "TLD extensions to check (e.g., ['com', 'io', 'dev']). Defaults to ['com', 'io', 'dev'].",
            },
            registrars: {
                type: 'array',
                items: { type: 'string' },
                description: "Optional: specific registrars to check. Leave empty to auto-select.",
            },
        },
        required: ['domain_name'],
    },
};
/**
 * Execute the search_domain tool.
 */
async function executeSearchDomain(input) {
    try {
        const { domain_name, tlds, registrars } = exports.searchDomainSchema.parse(input);
        return await (0, domain_search_js_1.searchDomain)(domain_name, tlds, registrars);
    }
    catch (error) {
        throw (0, errors_js_1.wrapError)(error);
    }
}
