// domain-search-mcp/dist/openapi/generator.js
"use strict";
/**
 * OpenAPI Specification Generator
 *
 * Generates OpenAPI 3.1 spec from MCP tool definitions.
 * Enables ChatGPT Actions and other REST API consumers.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateOpenAPISpec = generateOpenAPISpec;
exports.getOpenAPISpecJson = getOpenAPISpecJson;
const zod_to_openapi_1 = require("@asteasolutions/zod-to-openapi");
const zod_1 = require("zod");
// Extend Zod with OpenAPI methods
(0, zod_to_openapi_1.extendZodWithOpenApi)(zod_1.z);
// Import all tool schemas
const index_js_1 = require("../tools/index.js");
// Map tools to their schemas
const TOOL_CONFIGS = [
    {
        name: index_js_1.searchDomainTool.name,
        description: index_js_1.searchDomainTool.description,
        schema: index_js_1.searchDomainSchema,
        operationId: 'searchDomain',
    },
    {
        name: index_js_1.bulkSearchTool.name,
        description: index_js_1.bulkSearchTool.description,
        schema: index_js_1.bulkSearchSchema,
        operationId: 'bulkSearch',
    },
    {
        name: index_js_1.compareRegistrarsTool.name,
        description: index_js_1.compareRegistrarsTool.description,
        schema: index_js_1.compareRegistrarsSchema,
        operationId: 'compareRegistrars',
    },
    {
        name: index_js_1.suggestDomainsTool.name,
        description: index_js_1.suggestDomainsTool.description,
        schema: index_js_1.suggestDomainsSchema,
        operationId: 'suggestDomains',
    },
    {
        name: index_js_1.suggestDomainsSmartTool.name,
        description: index_js_1.suggestDomainsSmartTool.description,
        schema: index_js_1.suggestDomainsSmartSchema,
        operationId: 'suggestDomainsSmart',
    },
    {
        name: index_js_1.tldInfoTool.name,
        description: index_js_1.tldInfoTool.description,
        schema: index_js_1.tldInfoSchema,
        operationId: 'getTldInfo',
    },
    {
        name: index_js_1.checkSocialsTool.name,
        description: index_js_1.checkSocialsTool.description,
        schema: index_js_1.checkSocialsSchema,
        operationId: 'checkSocials',
    },
    {
        name: index_js_1.analyzeProjectTool.name,
        description: index_js_1.analyzeProjectTool.description,
        schema: index_js_1.analyzeProjectSchema,
        operationId: 'analyzeProject',
    },
    {
        name: index_js_1.huntDomainsTool.name,
        description: index_js_1.huntDomainsTool.description,
        schema: index_js_1.huntDomainsSchema,
        operationId: 'huntDomains',
    },
    {
        name: index_js_1.expiringDomainsTool.name,
        description: index_js_1.expiringDomainsTool.description,
        schema: index_js_1.expiringDomainsSchema,
        operationId: 'getExpiringDomains',
    },
];
// Generic success response schema
const SuccessResponseSchema = zod_1.z.object({
    success: zod_1.z.boolean().describe('Whether the operation succeeded'),
    data: zod_1.z.unknown().describe('Tool-specific response data'),
});
// Error response schema
const ErrorResponseSchema = zod_1.z.object({
    error: zod_1.z.object({
        code: zod_1.z.string().describe('Error code'),
        message: zod_1.z.string().describe('Human-readable error message'),
        retryable: zod_1.z.boolean().optional().describe('Whether the request can be retried'),
    }),
});
/**
 * Generate OpenAPI specification from MCP tool definitions.
 *
 * @param baseUrl - Base URL for the API (e.g., https://api.example.com)
 * @param version - API version (defaults to package version)
 */
function generateOpenAPISpec(baseUrl = 'https://api.domain-search-mcp.com', version) {
    const registry = new zod_to_openapi_1.OpenAPIRegistry();
    // Register response schemas
    registry.register('SuccessResponse', SuccessResponseSchema);
    registry.register('ErrorResponse', ErrorResponseSchema);
    // Register each tool as an API endpoint
    for (const tool of TOOL_CONFIGS) {
        // ChatGPT has a 300 char limit for operation descriptions
        const truncatedDescription = tool.description.length > 295
            ? tool.description.slice(0, 292) + '...'
            : tool.description;
        registry.registerPath({
            method: 'post',
            path: `/api/tools/${tool.name}`,
            operationId: tool.operationId,
            summary: tool.name.replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()),
            description: truncatedDescription,
            request: {
                body: {
                    content: {
                        'application/json': {
                            schema: tool.schema,
                        },
                    },
                    required: true,
                },
            },
            responses: {
                200: {
                    description: 'Successful response',
                    content: {
                        'application/json': {
                            schema: SuccessResponseSchema,
                        },
                    },
                },
                400: {
                    description: 'Bad request - invalid parameters',
                    content: {
                        'application/json': {
                            schema: ErrorResponseSchema,
                        },
                    },
                },
                429: {
                    description: 'Rate limit exceeded',
                    content: {
                        'application/json': {
                            schema: ErrorResponseSchema,
                        },
                    },
                },
                500: {
                    description: 'Internal server error',
                    content: {
                        'application/json': {
                            schema: ErrorResponseSchema,
                        },
                    },
                },
            },
        });
    }
    // Generate the OpenAPI document
    const generator = new zod_to_openapi_1.OpenApiGeneratorV31(registry.definitions);
    const apiVersion = version || require('../../package.json').version;
    return generator.generateDocument({
        openapi: '3.1.0',
        info: {
            title: 'Domain Search MCP API',
            version: apiVersion,
            description: `REST API for domain search operations. Provides domain availability checking, AI-powered suggestions, pricing comparison, and social handle validation.

This API is auto-generated from MCP tool definitions and can be used with:
- ChatGPT Actions / Custom GPTs
- OpenAI Assistants
- Any REST API client

**Note:** For MCP protocol clients (Claude Desktop, Cursor, VS Code), use the MCP endpoint at \`/mcp\` instead.`,
            contact: {
                name: 'Domain Search MCP',
                url: 'https://github.com/dorukardahan/domain-search-mcp',
            },
            license: {
                name: 'MIT',
                url: 'https://opensource.org/licenses/MIT',
            },
        },
        servers: [
            {
                url: baseUrl,
                description: 'Production server',
            },
        ],
        tags: [
            {
                name: 'domains',
                description: 'Domain availability and search operations',
            },
            {
                name: 'suggestions',
                description: 'AI-powered domain name suggestions',
            },
            {
                name: 'social',
                description: 'Social media handle availability',
            },
        ],
    });
}
/**
 * Get the OpenAPI spec as a JSON string.
 */
function getOpenAPISpecJson(baseUrl, version) {
    return JSON.stringify(generateOpenAPISpec(baseUrl, version), null, 2);
}
