// domain-search-mcp/dist/utils/concurrency.js
"use strict";
/**
 * Concurrency limiters for async work.
 *
 * Keeps upstream services from being flooded and helps control latency.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyedLimiter = exports.ConcurrencyLimiter = void 0;
class ConcurrencyLimiter {
    limit;
    active = 0;
    queue = [];
    constructor(limit) {
        this.limit = limit;
    }
    async run(fn) {
        await this.acquire();
        try {
            return await fn();
        }
        finally {
            this.release();
        }
    }
    acquire() {
        if (this.active < this.limit) {
            this.active += 1;
            return Promise.resolve();
        }
        return new Promise((resolve) => {
            this.queue.push(() => {
                this.active += 1;
                resolve();
            });
        });
    }
    release() {
        this.active = Math.max(0, this.active - 1);
        const next = this.queue.shift();
        if (next)
            next();
    }
}
exports.ConcurrencyLimiter = ConcurrencyLimiter;
class KeyedLimiter {
    limit;
    limiters = new Map();
    constructor(limit) {
        this.limit = limit;
    }
    async run(key, fn) {
        let limiter = this.limiters.get(key);
        if (!limiter) {
            limiter = new ConcurrencyLimiter(this.limit);
            this.limiters.set(key, limiter);
        }
        return limiter.run(fn);
    }
}
exports.KeyedLimiter = KeyedLimiter;
