// domain-search-mcp/dist/registrars/index.js
"use strict";
/**
 * Registrar Exports.
 *
 * Porkbun is the primary registrar for availability and pricing checks.
 * Namecheap is supported as an alternative.
 * GoDaddy public endpoint provides free availability + premium/auction detection.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetGodaddyCircuit = exports.getGodaddyCircuitState = exports.godaddyPublicAdapter = exports.GodaddyPublicAdapter = exports.namecheapAdapter = exports.NamecheapAdapter = exports.porkbunAdapter = exports.PorkbunAdapter = exports.RateLimiter = exports.RegistrarAdapter = void 0;
var base_js_1 = require("./base.js");
Object.defineProperty(exports, "RegistrarAdapter", { enumerable: true, get: function () { return base_js_1.RegistrarAdapter; } });
Object.defineProperty(exports, "RateLimiter", { enumerable: true, get: function () { return base_js_1.RateLimiter; } });
var porkbun_js_1 = require("./porkbun.js");
Object.defineProperty(exports, "PorkbunAdapter", { enumerable: true, get: function () { return porkbun_js_1.PorkbunAdapter; } });
Object.defineProperty(exports, "porkbunAdapter", { enumerable: true, get: function () { return porkbun_js_1.porkbunAdapter; } });
var namecheap_js_1 = require("./namecheap.js");
Object.defineProperty(exports, "NamecheapAdapter", { enumerable: true, get: function () { return namecheap_js_1.NamecheapAdapter; } });
Object.defineProperty(exports, "namecheapAdapter", { enumerable: true, get: function () { return namecheap_js_1.namecheapAdapter; } });
var godaddy_public_js_1 = require("./godaddy-public.js");
Object.defineProperty(exports, "GodaddyPublicAdapter", { enumerable: true, get: function () { return godaddy_public_js_1.GodaddyPublicAdapter; } });
Object.defineProperty(exports, "godaddyPublicAdapter", { enumerable: true, get: function () { return godaddy_public_js_1.godaddyPublicAdapter; } });
Object.defineProperty(exports, "getGodaddyCircuitState", { enumerable: true, get: function () { return godaddy_public_js_1.getGodaddyCircuitState; } });
Object.defineProperty(exports, "resetGodaddyCircuit", { enumerable: true, get: function () { return godaddy_public_js_1.resetGodaddyCircuit; } });
