// domain-search-mcp/dist/aftermarket/sedo.js
"use strict";
/**
 * Sedo public auctions feed integration (no auth).
 *
 * Feed format (semicolon-separated):
 * domain;auction_start;auction_end;price;currency;...;
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseSedoLine = parseSedoLine;
exports.parseSedoFeed = parseSedoFeed;
exports.lookupSedoAuction = lookupSedoAuction;
const config_js_1 = require("../config.js");
const logger_js_1 = require("../utils/logger.js");
const cache_js_1 = require("../utils/cache.js");
const FEED_CACHE_KEY = 'sedo:auctions';
const feedCache = new cache_js_1.TtlCache(config_js_1.config.cache.sedoTtl, 2);
function normalizeCurrency(raw) {
    if (!raw)
        return null;
    const cleaned = raw.replace(/[^A-Za-z]/g, '').toUpperCase();
    if (cleaned === 'US')
        return 'USD';
    if (cleaned.length === 3)
        return cleaned;
    return null;
}
function buildSedoSearchUrl(domain) {
    return `https://sedo.com/search/?keyword=${encodeURIComponent(domain)}`;
}
function parseSedoLine(line) {
    const trimmed = line.trim();
    if (!trimmed)
        return null;
    const parts = trimmed.split(';');
    if (parts.length < 5)
        return null;
    const domain = parts[0]?.trim().toLowerCase();
    if (!domain)
        return null;
    const priceRaw = parts[3]?.trim();
    const price = priceRaw ? Number(priceRaw) : NaN;
    const currency = normalizeCurrency(parts[4]?.trim() || '');
    const endTs = parts[2] ? Number(parts[2]) : NaN;
    const auctionEnd = Number.isFinite(endTs)
        ? new Date(endTs * 1000).toISOString()
        : null;
    return {
        domain,
        price: Number.isFinite(price) ? price : null,
        currency,
        auction_end: auctionEnd,
        source: 'sedo_feed',
        url: buildSedoSearchUrl(domain),
    };
}
function parseSedoFeed(text) {
    const index = new Map();
    for (const line of text.split('\n')) {
        const listing = parseSedoLine(line);
        if (listing) {
            index.set(listing.domain, listing);
        }
    }
    return index;
}
async function fetchSedoFeed() {
    const cached = feedCache.get(FEED_CACHE_KEY);
    if (cached)
        return cached;
    const url = config_js_1.config.aftermarket.sedoFeedUrl || 'https://sedo.com/txt/auctions_us.txt';
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5000);
    try {
        const response = await fetch(url, { signal: controller.signal });
        if (!response.ok) {
            throw new Error(`Sedo feed HTTP ${response.status}`);
        }
        const text = await response.text();
        const index = parseSedoFeed(text);
        feedCache.set(FEED_CACHE_KEY, index);
        return index;
    }
    finally {
        clearTimeout(timeout);
    }
}
async function lookupSedoAuction(domain) {
    if (!config_js_1.config.aftermarket.sedoEnabled)
        return null;
    try {
        const index = await fetchSedoFeed();
        return index.get(domain.toLowerCase()) ?? null;
    }
    catch (error) {
        logger_js_1.logger.debug('Sedo feed lookup failed', {
            domain,
            error: error instanceof Error ? error.message : String(error),
        });
        return null;
    }
}
